ompi/util/count_disp_array.h

```diff
@@ -0,0 +1,136 @@
+        } \
+    } while (0)
+#endif
+
+/* Return if the internal type is 64-bit or not */
+static inline bool ompi_count_array_is_64bit(ompi_count_array_t *array)
+{
+    return !((uint64_t)array->data.size_t_array & 0x1L);
+}
+
+static inline const void *ompi_count_array_ptr(ompi_count_array_t *array)
+{
+    if (OPAL_LIKELY((uint64_t)array->data.int_array & 0x1L)){
+        return (const void *)((uint64_t)array->data.int_array & ~0x1L);
+    }
+    return array->data.size_t_array;
+}
+
+/* Get a count in the array at index i */
+static inline size_t ompi_count_array_get(ompi_count_array_t *array, size_t i)
+{
+    if (OPAL_LIKELY((uint64_t)array->data.int_array & 0x1L)){
+        const int *iptr = (const int *)((uint64_t)array->data.int_array & ~0x1L);
+        return iptr[i];
+    }
+    return array->data.size_t_array[i];
+}
+
+typedef struct ompi_disp_array_t {
+    union {
+        const int *int_array;
+        const ptrdiff_t *ptrdiff_t_array;
+    } data;
+} ompi_disp_array_t;
+
+/* Initialize an int variant of the disp array */
+static inline void ompi_disp_array_init(ompi_disp_array_t *array, const int *data)
+{
+    array->data.int_array = (const int *)((uint64_t)data | 0x1L);
+}
+
+/* Initialize a bigcount variant of the disp array */
+static inline void ompi_disp_array_init_c(ompi_disp_array_t *array, const ptrdiff_t *data)
+{
+    array->data.ptrdiff_t_array = data;
+}
+
+#if OPAL_C_HAVE__GENERIC
+#define OMPI_DISP_ARRAY_INIT(array, data) _Generic((data), \
+                                                   int *: ompi_disp_array_init, \
+                                                   const int *: ompi_disp_array_init, \
+                                                   ptrdiff_t *: ompi_disp_array_init_c, \
+                                                   const ptrdiff_t *: ompi_disp_array_init_c)(array, data)
+#else
+#define OMPI_DISP_ARRAY_INIT(array, data) \
+    do { \
+        if (sizeof(*(data)) == sizeof(int)) { \
+            ompi_disp_array_init(array, (const int *) (data)); \
+        } else if (sizeof(*(data)) == sizeof(ptrdiff_t)) { \
+            ompi_disp_array_init_c(array, (const ptrdiff_t *) (data)); \
+        } \
+    } while(0)
+#endif
+
+/* Return if the internal type is 64-bit or not */
+static inline bool ompi_disp_array_is_64bit(ompi_disp_array_t *array)
+{
+    return !((uint64_t)array->data.ptrdiff_t_array & 0x1L);
+}
+
+/* Get a displacement in the array at index i */
+static inline ptrdiff_t ompi_disp_array_get(ompi_disp_array_t *array, size_t i)
+{
+    if (OPAL_LIKELY((uint64_t)array->data.int_array & 0x1L)){
+        const int *iptr = (const int *)((uint64_t)array->data.int_array & ~0x1L);
+        return iptr[i];
+    }
+    return array->data.ptrdiff_t_array[i];
+}
+
+static inline const void *ompi_disp_array_ptr(ompi_disp_array_t *array)
+{
+    if (OPAL_LIKELY((uint64_t)array->data.int_array & 0x1L)){
+        return (const void *)((uint64_t)array->data.int_array & ~0x1L);
```

Comment:
Just a small suggestion: I'd prefer `intptr_t` instead of `uint64_t`. `long` (as in `~0x1L`) is not guaranteed to be 64bit on all machines but the promotion to (signed) `intptr_t` in two-compliment will guarantee that the sign is propagated and we end up with a correct 64bit mask. I'm not sure this will work with unsigned 64bit. Also, `intptr_t` is guaranteed to be large enough to hold a pointer.